packages/grafana-sql/src/datasource/SqlDatasource.ts

```diff
@@ -67,6 +82,173 @@ export abstract class SqlDatasource extends DataSourceWithBackend<SQLQuery, SQLO
+      if (orinQuery) {
+        sqlExpress.push(this.prepareSqlExpress(duplicate, row, orinQuery))
+      }
+    }
+
+    const contextRequest: DataQueryRequest<SQLQuery> = {
+      requestId: `mysql-log-context-${row.dataFrame.refId}`,
+      targets: sqlExpress,
+      interval: interval.interval,
+      intervalMs: interval.intervalMs,
+      range,
+      scopedVars: scopedVars,
+      timezone: 'UTC',
+      app: CoreApp.Explore,
+      startTime: Date.now(),
+      hideFromInspector: true,
+    };
+    return contextRequest;
+  }
+
+  prepareScopeVars(duplicate: Labels) {
+    const scopedVars: ScopedVars = {}
+    this.templateSrv.getVariables().map(v => {
+      if (v.name in duplicate) {
+        scopedVars[v.name] = { text: `'${duplicate[v.name]}'`, value: `'${duplicate[v.name]}'` };
+        delete duplicate[v.name];
+      }
+    });
+    return scopedVars;
+  }
+
+  prepareSqlExpress(duplicate: Labels, row: LogRowModel, orinQuery: SQLQuery): SQLQuery {
+    const query = JSON.parse(JSON.stringify(orinQuery));
+    let whereClause = 'WHERE ';
+    const whereLiteral = this.getWhereLiteral(query.rawSql || "")
+    if (whereLiteral.length > 0) {
+      query.rawSql = query.rawSql?.replace(whereLiteral, whereClause);
+    } else {
+      const table = this.getTablename(query.rawSql || "");
+      if (table.length > 0) {
+        whereClause = table + whereClause + "1=1";
+        query.rawSql = query.rawSql?.replace(table, whereClause);
+      }
+    }
+    return query;
+  }
+
+  processOrderByClause(sql: string, keyword: 'ASC' | 'DESC'): string {
+    const orderByRegex = /order\s+by\s+([\w.,\s]+?)(\s+asc|\s+desc)?(\s+|$)/i;
+    if (!orderByRegex.test(sql)) {
+      return sql; // not exist order by clause
+    }
+    return sql.replace(orderByRegex, (_, columns) => `ORDER BY ${columns.trim()} ${keyword} `);
+  }
+
+  getWhereLiteral(sql: string): string {
+    const regex = /\bwhere\b/i;
+    const match = sql.match(regex);
+    if (match) {
+      return match[0];
+    }
+    return "";
+  }
+
+  getTablename(sql: string): string {
+    const regex = /from\s+(\S+)/i;
+    const match = sql.match(regex);
+    if (match) {
+      return match[1];
+    }
+    return "";
+  }
+
+  // Acquire the log rows context from sql dataSource
+  getLogRowContext(row: LogRowModel, options?: LogRowContextOptions, query?: SQLQuery): Promise<DataQueryResponse> {
+    const contextRequest: DataQueryRequest<SQLQuery> = this.makeLogContextDataRequest(row, options, query);
+    return lastValueFrom(
+      this.query(contextRequest).pipe(
+        catchError((err) => {
+          const error: DataQueryError = {
+            message: 'Error during context query. Please check JS console logs.',
+            status: err.status,
+            statusText: err.statusText,
+          };
+          throw error;
+        })
+      )
+    );
+  }
+
+  getLogRowContextUi(row: LogRowModel, runContextQuery?: () => void, orignQuery?: SQLQuery): React.ReactNode {
```

Comment:
This API is currently an @alpha API is currently under development. @grafana/observability-logs what are your thoughts on potentially adding this api to sql data sources?

Reply:
I think it makes sense, as long as they produce log data frames. Do you think there could be any hidden gotcha? SQL data sources could take advantage of the table visualization.